nni/retiarii/strategy/bruteforce.py

```diff
@@ -0,0 +1,95 @@
+
+
+def random_generator(search_space: Dict[Any, List[Any]], dedup=True, retries=500):
+    keys = list(search_space.keys())
+    history = set()
+    search_space_values = copy.deepcopy(list(search_space.values()))
+    while True:
+        for retry_count in range(retries):
+            selected = [random.choice(v) for v in search_space_values]
+            if not dedup:
+                break
+            selected = tuple(selected)
+            if selected not in history:
+                history.add(selected)
+                break
+            if retry_count + 1 == retries:
+                _logger.info('Random generation has run out of patience. There is nothing to search. Exiting.')
+                return
+        yield {key: value for key, value in zip(keys, selected)}
+
+
+class GridSearch(BaseStrategy):
+    def __init__(self, shuffle=True):
+        self._polling_interval = 2.
+        self.shuffle = shuffle
+
+    def run(self, base_model, applied_mutators):
+        search_space = dry_run_for_search_space(base_model, applied_mutators)
+        for sample in grid_generator(search_space, shuffle=self.shuffle):
+            _logger.info('New model created. Waiting for resource. %s', str(sample))
+            if query_available_resources() <= 0:
+                time.sleep(self._polling_interval)
+            submit_models(get_targeted_model(base_model, applied_mutators, sample))
+
+
+class _RandomSampler(Sampler):
+    def choice(self, candidates, mutator, model, index):
+        return random.choice(candidates)
+
+
+class Random(BaseStrategy):
+    def __init__(self, variational=False, dedup=True):
+        self.variational = variational
+        self.dedup = dedup
+        if variational and dedup:
+            raise ValueError('Dedup is not supported in variational mode.')
+        self.random_sampler = _RandomSampler()
+        self._polling_interval = 2.
+
+    def run(self, base_model, applied_mutators):
+        if self.variational:
+            _logger.info('Random search running in variational mode.')
+            sampler = _RandomSampler()
+            for mutator in applied_mutators:
+                mutator.bind_sampler(sampler)
+            while True:
+                avail_resource = query_available_resources()
+                if avail_resource > 0:
+                    model = base_model
+                    for mutator in applied_mutators:
+                        model = mutator.apply(model)
+                    _logger.info('New model created. Applied mutators are: %s', str(applied_mutators))
+                    submit_models(model)
+                else:
+                    time.sleep(self._polling_interval)
+        else:
+            _logger.info('Random search running in fixed size mode. Dedup: %s.', 'on' if self.dedup else 'off')
+            search_space = dry_run_for_search_space(base_model, applied_mutators)
+            for sample in random_generator(search_space, dedup=self.dedup):
```

Comment:
what will webui show if strategy exits but maxtrialnum and maxduration are not reached?

Reply:
I tested. The whole experiment directly exits and dispatcher (strategy) is terminated.

Reply:
dispatcher is not strategy, why dispatcher exits?...

Reply:
"Dispatcher terminated" is printed on the console. Afterwards, no more trials appear on the WebUI. I'm not sure about the details.

Reply:
got it, this is a normal behavior